contracts/erc20-bridge-sampler/contracts/src/ERC20BridgeSampler.sol

```diff
@@ -639,4 +680,133 @@ contract ERC20BridgeSampler is
+                takerToken,
+                makerToken,
+                tmpTakerAmounts
+            );
+        }
+        (bool success, bytes memory resultData) = address(this).staticcall(callData);
+        if (!success) {
+            return 0;
+        }
+        // solhint-disable indent
+        takerTokenAmount = abi.decode(resultData, (uint256[]))[0];
+    }
+
+    function _sampleApproximateBuysFromSource(
+        address takerToken,
+        address makerToken,
+        uint256[] memory makerTokenAmounts,
+        FakeBuyOptions memory opts,
+        bytes4 selector,
+        address plpRegistryAddress
+    )
+        private
+        view
+        returns (uint256[] memory takerTokenAmounts)
+    {
+        _assertValidPair(makerToken, takerToken);
+        if (makerTokenAmounts.length == 0) {
+            return takerTokenAmounts;
+        }
+        uint256 sellAmount;
+        uint256 buyAmount;
+        uint256 slippageFromTarget;
+        takerTokenAmounts = new uint256[](makerTokenAmounts.length);
+        sellAmount = _sampleSellForApproximateBuy(
+            makerToken,
+            takerToken,
+            makerTokenAmounts[0],
+            selector,
+            plpRegistryAddress
+        );
+
+        if (sellAmount == 0) {
+            return takerTokenAmounts;
+        }
+
+        buyAmount = _sampleSellForApproximateBuy(
+            takerToken,
+            makerToken,
+            sellAmount,
+            selector,
+            plpRegistryAddress
+        );
+        if (buyAmount == 0) {
+            return takerTokenAmounts;
+        }
+
+        for (uint256 i = 0; i < makerTokenAmounts.length; i++) {
+            uint256 iteration = 0;
+            // Default to some value higher than our target
+            slippageFromTarget = opts.targetSlippageBps + 1;
+            do {
+                // adjustedSellAmount = previousSellAmount * (target/actual) * JUMP_MULTIPLIER
+                sellAmount = LibMath.getPartialAmountCeil(
+                    makerTokenAmounts[i],
+                    buyAmount,
+                    sellAmount
+                );
+                // JUMP Multiplier by 1.0005
+                sellAmount = LibMath.getPartialAmountCeil(
+                    (10000 + opts.targetSlippageBps),
+                    10000,
+                    sellAmount
+                );
+                buyAmount = _sampleSellForApproximateBuy(
+                    takerToken,
+                    makerToken,
+                    sellAmount,
+                    selector,
+                    plpRegistryAddress
+                );
+                if (buyAmount == 0) {
+                    break;
+                }
+
+                // 0.0005 slippage is the target
+                if (buyAmount >= makerTokenAmounts[i]) {
+                    slippageFromTarget = (buyAmount - makerTokenAmounts[i]) /
+                                        (makerTokenAmounts[i] / 10000);
+                }
```

Comment:
I feel like the slippage break should be in here. I think it's clearer and less error-prone since the `slippageFromTarget` is only guaranteed to be up-to-date within this block. 

So like
```solidity
  if (buyAmount >= makerTokenAmounts[i]) {
    slippageFromTarget = ...
    if (slippageFromTarget <= opts.targetSlippageBps) break;
  }
} while (++iteration < opts.maxIterations);
```

From here we can maybe turn the do...while into a more conventional for-loop `for(iteration = 0; iteration < opts.maxIterations; ++iteration)`